source/agora/consensus/protocol/Nominator.d

```diff
@@ -1081,6 +1096,123 @@ extern(D):
+
+        Get the UTXO hash for the index of the validator at a height
+
+        Params:
+            slot_idx = the height for which we search the index of the validator
+            index = the index of the validator
+
+        Returns:
+            the hash of the frozen UTXO
+
+    ***************************************************************************/
+
+    public StellarHash getNodeUTXO (uint64_t slot_index, uint64_t index)
+        @safe nothrow
+    {
+        Hash utxo;
+        if (slot_index == 1)
+        {
+            Block b0;
+            try
+            {
+                b0 = this.ledger.getBlocksFrom(Height(0))[0];
+            }
+            catch (Exception exc)
+            {
+                log.error("Exception happened with getting block at requested height: {}", exc);
+            }
+            assert(b0.header.enrollments.length > index);
+            utxo = b0.header.enrollments[index].utxo_key;
+        }
+        else
+        {
+            ValidatorInfo[] validators;
+            try
+            {
+                validators = this.ledger.getValidators(Height(slot_index));
+            }
+            catch (Exception exc)
+            {
+                log.error("Exception happened with calling `getValidators` in `getNodeID: {}", exc);
+            }
+
+            if (validators.length > index)
+                utxo = validators[index].utxo;
+        }
+        if (utxo == Hash.init)
+        {
+            log.trace("utxo is Hash.init", slot_index);
+            assert(0, "utxo is Hash.init");
+        }
+
+        return StellarHash(utxo[][0 .. StellarHash.sizeof]);
+    }
+
+    /***************************************************************************
+
+        Get the index of this validator at a height.
+
+        Params:
+            slot_idx = the height for which we search the index of the validator
+
+        Returns:
+            the index of the validator
+
+    ***************************************************************************/
+
+    public uint64_t getSelfUTXOIndex (uint64_t slot_index) @safe nothrow
+    {
+        auto self_enroll = this.enroll_man.getEnrollmentKey();
+        if (self_enroll == Hash.init)
+        {
+            return uint64_t.max;
+        }
+
+        Block block;
+        try
+        {
+            block = this.ledger.getBlocksFrom(Height(0))[0];
+        }
+        catch (Exception exc)
+        {
+            log.fatal("Exception happened with calling `getBlocksFrom` on `getNodeIndex: {}", exc);
+            assert(0);
+        }
+
+        const node_hash = StellarHash(self_enroll[][0 .. StellarHash.sizeof]);
+        if (slot_index == 0)
+        {
+            foreach (idx, enroll; block.header.enrollments)
+            {
+                if (enroll.utxo_key == node_hash)
+                    return idx;
+            }
+            log.fatal("Could not find a UTXO for this node at Genesis");
+            assert(0);
+        }
```

Comment:
The only call I can see is at L243 in validator, which means we will never call this with `slot_index == 0`.

Reply:
This is also called at L160 in `Validator.d`.

Reply:
Ditto